src/vanilla/shallow.ts

```diff
@@ -1,6 +1,9 @@
 const isIterable = (obj: object): obj is Iterable<unknown> =>
   Symbol.iterator in obj
 
+const type = <T>(obj: T): string =>
+  Object.prototype.toString.apply(obj).slice(8, -1).toLowerCase()
+
```

Comment:
I don't think this is very nice and acceptable. If we want to distinguish an array, use `Array.isArray()`.

Reply:
Here's one solution:

```diff
diff --git a/src/vanilla/shallow.ts b/src/vanilla/shallow.ts
index 749e99b..09f594b 100644
--- a/src/vanilla/shallow.ts
+++ b/src/vanilla/shallow.ts
@@ -57,6 +57,9 @@ export function shallow<T>(valueA: T, valueB: T): boolean {
   ) {
     return false
   }
+  if (Array.isArray(valueA) !== Array.isArray(valueB)) {
+    return false
+  }
   if (!isIterable(valueA) || !isIterable(valueB)) {
     return compareEntries(
       { entries: () => Object.entries(valueA) },
```

Reply:
https://github.com/pmndrs/zustand/pull/3108#discussion_r2075038796